contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -891,12 +891,22 @@ template<class t_connection_context> template<class callback_t>
 bool async_protocol_handler_config<t_connection_context>::foreach_connection(const callback_t &cb)
 {
   CRITICAL_REGION_LOCAL(m_connects_lock);
-  for(auto& c: m_connects)
-  {
-    async_protocol_handler<t_connection_context>* aph = c.second;
-    if(!cb(aph->get_context_ref()))
+  std::vector<typename connections_map::mapped_type> conn;
+  conn.reserve(m_connects.size());
+
+  auto scope_exit_handler = misc_utils::create_scope_leave_handler([&conn]{
```

Comment:
Good grief.

I'm being lazy (at the moment) by not running this locally, but `start_outer_call()` and `finish_outer_call()` shouldn't be needed in the change. The issue is that you cannot modify `m_connects` while iterating over it. The lock protects against other threads, but it is a recursive lock so the callback can interfere with the container.

```c++
std::vector<typename connections_map::mapped_type> conn{m_connects};
for (auto &aph : conn)
```
should be sufficient and incur less CPU penalty.

The `start_outer_call()` isn't needed because `m_connects_lock` protects other threads from removing the connection, and if the callback tries to destroy the connection the destructor waits for 60s before blowing up. Which means the callback cannot destroy the connection in this design.

Reply:
Or wait, copying the `.second` requires more than a constructor call but I think the idea is clear.

Reply:
Disregard.